src/extension_commands.ml

```diff
@@ -766,6 +766,121 @@ module MerlinJump = struct
     command Extension_consts.Commands.merlin_jump handler
 end
 
+module Navigate_holes = struct
+  let extension_name = "Navigate between Typed Holes"
+
+  let ocaml_lsp_doesnt_support_typed_holes ocaml_lsp =
+    not (Ocaml_lsp.can_handle_typed_holes ocaml_lsp)
+
+  let is_valid_text_doc textdoc =
+    match TextDocument.languageId textdoc with
+    | "ocaml" | "ocaml.interface" | "reason" -> true
+    | _ -> false
+
+  let send_request_to_lsp client doc =
+    let uri = TextDocument.uri doc in
+    Custom_requests.send_request client Custom_requests.typedHoles uri
+
+  let display_results (results : Range.t list) text_document =
+    let quickPickItems =
+      List.map results ~f:(fun res ->
+          let line = Position.line @@ Range.end_ res in
+          ( QuickPickItem.create
+              ~label:
+                (Printf.sprintf
+                   "Line %d: %s"
+                   line
+                   (TextLine.text @@ TextDocument.lineAt ~line text_document))
+              ()
+          , (res, ()) ))
+    in
+    let quickPickOptions = QuickPickOptions.create ~title:"Typed Holes" () in
+    Window.showQuickPickItems
+      ~choices:quickPickItems
+      ~options:quickPickOptions
+      ()
+
+  let jump_to_hole range text_editor =
+    let open Promise.Syntax in
+    let+ _ =
+      Window.showTextDocument
+        ~document:(TextEditor.document text_editor)
+        ~preserveFocus:true
+        ()
+    in
+    let new_selection =
+      let anchor = Range.start range in
+      let active = Range.end_ range in
+      Selection.makePositions ~anchor ~active
+    in
+    TextEditor.set_selection text_editor new_selection;
+    TextEditor.revealRange
+      text_editor
+      ~range
+      ~revealType:TextEditorRevealType.InCenterIfOutsideViewport
+      ()
+
+  let handle_hole_navigation text_editor client instance =
+    let open Promise.Syntax in
+    let doc = TextEditor.document text_editor in
+    let* hole_positions = send_request_to_lsp client doc in
+    match hole_positions with
+    | [] ->
+      show_message `Info "No typed holes found in the file.";
+      Promise.return ()
+    | holes -> (
+      let* selected_hole = display_results holes doc in
+      match selected_hole with
+      | Some (range, ()) -> (
+        let* _ = jump_to_hole range text_editor in
+        match
+          Settings.(get server_typedHolesConstructAfterNavigate_setting)
+        with
+        | Some true | None ->
+          Construct.process_construct
+            (Range.end_ range)
+            text_editor
+            client
+            instance
+        | Some false -> Promise.return ())
+      | None -> Promise.return ())
+
+  let _holes =
+    let handler (instance : Extension_instance.t) ~args:_ =
+      let hole_navigation () =
```

Comment:
Since `hole_navigation` returns `unit` and is called right after, I don't think there is any benefit in giving it a name.
```suggestion
```